lib/src/node/value/calculation.dart

```diff
@@ -0,0 +1,144 @@
+// Copyright 2021 Google Inc. Use of this source code is governed by an
+// MIT-style license that can be found in the LICENSE file or at
+// https://opensource.org/licenses/MIT.
+
+import 'dart:js_util';
+
+import 'package:collection/collection.dart';
+import 'package:node_interop/js.dart';
+import 'package:sass/src/node/immutable.dart';
+import 'package:sass/src/node/utils.dart';
+
+import '../../value.dart';
+import '../reflection.dart';
+
+/// Check that [arg] is a valid argument to a calculation function.
+void isCalculationValue(Object arg, {bool checkUnquoted = true}) {
+  if (arg is! SassNumber &&
+      arg is! SassString &&
+      arg is! SassCalculation &&
+      arg is! CalculationOperation &&
+      arg is! CalculationInterpolation) {
+    jsThrow(JsError('Argument must be one of '
+        'SassNumber, SassString, SassCalculation, CalculationOperation, '
+        'CalculationInterpolation'));
+  }
+  if (checkUnquoted && arg is SassString && arg.hasQuotes) {
+    jsThrow(JsError('Argument must be unquoted SassString'));
+  }
+}
+
+/// The JavaScript `SassCalculation` class.
+final JSClass calculationClass = () {
+  calc(Object argument) {
+    isCalculationValue(argument);
+    return SassCalculation.unsimplified('calc', [argument]);
+  }
+
+  min(Object arguments) {
+    var argList = jsToDartList(arguments).cast<Object>();
+    argList.forEach(isCalculationValue);
+    return SassCalculation.unsimplified('min', argList);
+  }
+
+  max(Object arguments) {
+    var argList = jsToDartList(arguments).cast<Object>();
+    argList.forEach(isCalculationValue);
+    return SassCalculation.unsimplified('max', argList);
+  }
+
+  clamp(Object min, [Object? value, Object? max]) {
+    if (value == null && max != null) {
+      jsThrow(JsError('`value` is undefined and `max` is not undefined'));
+    }
+    if ((value == null || max == null) &&
+        !(min is SassString && min.text.startsWith('var(')) &&
+        !(value is SassString && value.text.startsWith('var('))) {
+      jsThrow(JsError(
+          '`value` or `max` is undefined and neither `min` nor `value` is a SassString that begins with "var("'));
+    }
+    [min, value, max].whereNotNull().forEach(isCalculationValue);
+    return SassCalculation.unsimplified(
+        'clamp', [min, value, max].whereNotNull());
+  }
+
+  var jsClass =
+      createJSClass('sass.SassCalculation', (Object self, [Object? _]) {
+    jsThrow(JsError("new sass.SassCalculation() isn't allowed"));
```

Comment:
From the dart definition, I understand that direct instantiation is not allowed. Is that correct?

Reply:
That's right.